icu4c/source/test/intltest/csdetest.cpp

```diff
@@ -839,3 +843,21 @@ void CharsetDetectionTest::Ticket6954Test() {
     TEST_ASSERT(strcmp(name1, "windows-1252")==0);
 #endif
 }
+
+
+// Ticket 21823 - Issue with Charset Detector for ill-formed input strings. 
+//                Its fix involves returning a failure based error code 
+//                (U_INVALID_CHAR_FOUND) incase no charsets appear to match the input data.
+void CharsetDetectionTest::Ticket21823Test() {
+    UErrorCode status = U_ZERO_ERROR;
+    std::string str = "\x80";
+    UCharsetDetector* csd = ucsdet_open(&status);
+
+    ucsdet_setText(csd, str.data(), str.length(), &status);
+    const UCharsetMatch* match = ucsdet_detect(csd, &status);
+
+    if (match == NULL)
+        TEST_ASSERT(U_FAILURE(status));
```

Comment:
```suggestion
    if (match == NULL) {
        TEST_ASSERT(U_FAILURE(status));
    }
```

Reply:
> should we keep the current behavior of returning a NULL UCharsetMatch with no error code?

We could, although it's not great for ease-of-use. User code would need to check both the error code and nullptr.

I think changing the implementation so that the ISO 8859-1 / Windows 1252 detector matches anything, with a low probability, would also be reasonable. Anything is valid. For short sequences that don't appear to match anything else, it's as good a guess as any. The place to make this change is probably in `CharsetRecog_8859_1::match()`, in csrsbcs.cpp:892

If we stick with setting an error code, would U_ILLEGAL_ARGUMENT_ERROR be better than U_INVALID_CHAR_FOUND? Neither is a perfect fit, but there is not an invalid character found. The input provided is to short or weird for meaningful detection. Returning to the original question, not being able to detect isn't really an error condition at all - a detector shouldn't choke on anything you throw at it.